magenta-lib/src/main/scala/magenta/input/resolver/DeploymentResolver.scala

```diff
@@ -1,38 +1,41 @@
 package magenta.input.resolver
 
+import cats.data.Validated.{Invalid, Valid}
+import cats.data.{NonEmptyList => NEL, Validated, ValidatedNel}
+import cats.syntax.cartesian._
+import cats.syntax.traverse._
+import cats.instances.list._
 import magenta.input.{ConfigError, Deployment, DeploymentOrTemplate, RiffRaffDeployConfig}
 
 object DeploymentResolver {
-  def resolve(config: RiffRaffDeployConfig): List[Either[ConfigError, Deployment]] = {
-    config.deployments.map { case (label, rawDeployment) =>
+  def resolve(config: RiffRaffDeployConfig): ValidatedNel[ConfigError, List[Deployment]] = {
+    config.deployments.traverseU[ValidatedNel[ConfigError, Deployment]] { case (label, rawDeployment) =>
       for {
-        templated <- applyTemplates(label, rawDeployment, config.templates).right
-        deployment <- resolveDeployment(label, templated, config.stacks, config.regions).right
-        _ <- validateDependencies(label, deployment, config.deployments).right
-      } yield deployment
+        templated <- applyTemplates(label, rawDeployment, config.templates)
+        deployment <- resolveDeployment(label, templated, config.stacks, config.regions)
+        validatedDeployment <- validateDependencies(label, deployment, config.deployments)
+      } yield validatedDeployment
     }
   }
 
   /**
     * Validates and resolves a templated deployment by merging its
     * deployment attributes with any globally defined properties.
     */
-  private[input] def resolveDeployment(label: String, templated: DeploymentOrTemplate, globalStacks: Option[List[String]], globalRegions: Option[List[String]]): Either[ConfigError, Deployment] = {
-    for {
-      deploymentType <- templated.`type`.toRight(ConfigError(label, "No type field provided")).right
-      stacks <- templated.stacks.orElse(globalStacks).toRight(ConfigError(label, "No stacks provided")).right
-      regions <- templated.regions.orElse(globalRegions).toRight(ConfigError(label, "No regions provided")).right
-    } yield {
+  private[input] def resolveDeployment(label: String, templated: DeploymentOrTemplate, globalStacks: Option[List[String]], globalRegions: Option[List[String]]): ValidatedNel[ConfigError, Deployment] = {
+    (Validated.fromOption(templated.`type`, NEL.of(ConfigError(label, "No type field provided"))) |@|
```

Comment:
I think a custom error type with it's own `Semigroup` might avoid the littering of `NEL.of`.